src/EFCore.MySql/Storage/ValueComparison/Internal/JsonValueComparer.cs

```diff
@@ -0,0 +1,16 @@
+﻿using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace Pomelo.EntityFrameworkCore.MySql.Storage.ValueComparison.Internal
+{
+    public class JsonValueComparer<T> : ValueComparer<T>
+    {
+        public JsonValueComparer(ValueConverter valueConverter)
+            : base(
+                (left, right) => object.Equals(valueConverter.ConvertToProvider(left), valueConverter.ConvertToProvider(right)),
+                value => value.GetHashCode(),
+                value => (T)valueConverter.ConvertFromProvider(valueConverter.ConvertToProvider(value)))
```

Comment:
The snapshot lambda is invoked for every property when the entity is tracked or is saved. Making it faster would result in noticeable perf improvements.

Reply:
It could be improved for _some_ scenarios as far as I can see:
* Newtonsoft.Json if `T` is a DOM class: Use [JTokenEqualityComparer](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JTokenEqualityComparer.htm)
* Any Json-Library if `T` is `string`: Just compare the strings as is.

This would however do nothing for the most common case, which should be to use POCOs.

Reply:
You could check for `ICloneable` implementation, `JToken` has `DeepClone` and I believe JsonDocument is immutable, so same value can be returned

Reply:
> `JToken` has `DeepClone` and I believe `JsonDocument` is immutable

Good points!

> You could check for `ICloneable` implementation

Theoretically true, though the `ICloneable` contract is traditionally not explicit in _how_ the object is being cloned. From the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.icloneable?view=netcore-3.1#notes-to-implementers):

> Because callers of `Clone()` cannot depend on the method performing a predictable cloning operation, we recommend that `ICloneable` not be implemented in public APIs.

Reply:
You could also look for `DeepClone`, as long as this is documented it will allow to implement it efficiently for POCOs

Reply:
> You could also look for `DeepClone()`, as long as this is documented it will allow to implement it efficiently for POCOs

Could be a better approach. As for `ICloneable`, we might make this opt-in/opt-out.

Reply:
While JsonDocument is immutable and so there's no need to actually snapshot or do anything beyond reference comparison, JsonElement is a struct, and so that's not really possible... We should consider not allowing to map JsonElement.

Reply:
I agree that it makes little sense to snapshot `JsonElement` with the default struct member by member copy in the way `JsonElement` has been implemented. If the `JsonElement._parent` field had a publicly available property, this wouldn't be a problem, since the parent would still be immutable. But for some reason (no idea why) it has no property.

Reply:
You're right (https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1140#issuecomment-691620821)